### PR TITLE
CLI v0.5, compatible to python3

### PR DESCRIPTION
**Pull request by peter-tank**
API service refactoring

* Introspection and multicall API method restored(server side apply)
+ Try import ssl for API server
+ Test mode for API client `src/bmsxmlrpc/client.py`
+ Try compatible to python3(curses and PyQt4 GUI not tested.)

* Mainly changes:
  * Message attachments faults detecting, dump base64 embeded contents to files allowed
  * Multi-line message acceptable in sending(cancel with `Ctrl+D`), and allow resent on 'Connection Error.'
  * Print out detailed API Error returned.
  * Update Contacts list cmd, bcz contact list API returns unEncoded lable
  * Message pull routing refine, mainlly access inbox messages by IDs
  * API connections `SOCKS5` `HTTP` proxy supported
* UIs
  * Shorten the user cmds inputs
  * Try to replay user last choices.(rest to default selection by input a blank string)
  * A comprehensive command line parser, override configurations read from file `client.dat` in current working directory.(parse API settings from "keys.dat" either, if located in BMs directory)
  * Message preview limited to 380 characters in default and rewinds enabled

default settings `client.dat`
```
[global]
#start_daemon = http://127.0.0.1:8442

[api]
path = 127.0.0.1:8442
type = HTTP

[proxy]
path = 127.0.0.1:1080
type = none
timeout = 30
remotedns = True
```

### CODE REVIEW

File: src/depends.py [hunk 1]
```diff
@@ -137,6 +137,7 @@ def detectOSRelease():
         for line in osRelease:
             if line.startswith("NAME="):
                 detectOS.result = OS_RELEASE.get(
+                    line.replace('"', '').split("=")[-1].strip().lower().split(" ")[0])
                     line.replace('"', '').split("=")[-1].strip().lower())
```
Comment (peter-tank) [inline src/depends.py]: duplicated here on merge conflict resolving. it'll take a bit longer on trying a selective mergeing workflow.

**Comment by g1itch** [comment]
You eager to rewrite all pybitmessage code. Why don't you start from 2to3 or six?

**Comment by peter-tank** [comment]
@g1itch That's not true.
I just want to keep the CLI working as I expected, including under python3.
As we know, most bundled python2 not packed `six` into it.
And you can see I definetlly not familiar with python coding either.

Now, I'm trying to import unitest stuf into to [these API codes...](https://github.com/peter-tank/bmsxmlrpc) I'm stucked into `not well-formed (invalid token): line 1, column 0` on [Travis huilding test](https://travis-ci.org/peter-tank/bmsxmlrpc/jobs/411366986) responses. That's too much work for me already.

File: src/bmconfigparser.py [hunk 2]
```diff
@@ -84,14 +84,17 @@ def get(self, section, option, raw=False, fallback=UNSET):
...
+                return self._get(section, self._convert_to_boolean, field)
+            return self.getboolean(section, field)
         except (ConfigParser.NoSectionError, ConfigParser.NoOptionError,
                 ValueError, AttributeError):
             return False
 
     def safeGetInt(self, section, field, default=0):
         try:
-            return self._get(section, int, field)  # getint(section, field)
+            print('%s = %d' % (field, int(self.get(section, field, fallback=default))))
```
Comment (omkar1117) [inline src/bmconfigparser.py]: better go with logger instead of print statement.

File: src/bmsxmlrpc/__init__.py [hunk 3]
```diff
@@ -0,0 +1,5 @@
+"""
```
Comment (omkar1117) [inline src/bmsxmlrpc/__init__.py]: instead of triple quotes could you please use comments

**Comment by peter-tank** [comment]
Problems on starting `src/bitmessagemain.py --daemon` under python3(@bytes + str).
Seem's a big problem, I'm not willing to fix them, hope someone continue from this PR.

**Comment by dimyme** [comment]
> Problems on starting `src/bitmessagemain.py --daemon` under python3(@bytes + str).
> Seem's a big problem, I'm not willing to fix them, hope someone continue from this PR.

porting pyBM to py3 ?  that is very much work, mate

**Pull request closed without merging**